src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,231 @@
+            // If the token is about to expire, trigger a notification.
+            check(&env.emails, conn)
+        })
+        .await
+        .map_err(|err| anyhow!(err.to_string()))?
+    }
+}
+
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens = find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    if expired_tokens.len() == MAX_ROWS as usize {
+        warn!("The maximum number of API tokens per query has been reached. More API tokens might be processed on the next run.");
+    }
+    for token in &expired_tokens {
+        if let Err(e) = handle_expiring_token(conn, token, emails) {
+            error!(?e, "Failed to handle expiring token");
+        }
+    }
+
+    Ok(())
+}
+
+fn handle_expiring_token(
+    conn: &mut PgConnection,
+    token: &ApiToken,
+    emails: &Emails,
+) -> Result<(), anyhow::Error> {
+    let user = User::find(conn, token.user_id)?;
+    let recipient = match user.email(conn)? {
+        Some(email) => email,
+        None => return Err(anyhow!("No address found")),
+    };
+    let email = ExpiryNotificationEmail {
+        name: &user.gh_login,
+        token_name: &token.name,
+        expiry_date: token.expired_at.unwrap().and_utc(),
+    };
+    emails.send(&recipient, email)?;
+    // Update the token to prevent duplicate notifications.
+    diesel::update(token)
+        .set(api_tokens::expiry_notification_at.eq(now.nullable()))
+        .execute(conn)?;
+    Ok(())
+}
+
+/// Find all tokens that are not revoked and will expire within the specified number of days.
+pub fn find_tokens_expiring_within_days(
+    conn: &mut PgConnection,
+    days_until_expiry: chrono::TimeDelta,
+) -> QueryResult<Vec<ApiToken>> {
+    api_tokens::table
+        .filter(api_tokens::revoked.eq(false))
+        .filter(
+            api_tokens::expired_at
+                .is_not_null()
+                .and(api_tokens::expired_at.assume_not_null().gt(now)) // Ignore already expired tokens
+                .and(
+                    api_tokens::expired_at
+                        .assume_not_null()
+                        .lt(now + days_until_expiry.num_days().day()),
+                ),
+        )
+        .filter(api_tokens::expiry_notification_at.is_null())
+        .select(ApiToken::as_select())
+        .order_by(api_tokens::expired_at.asc()) // The most urgent tokens first
+        .limit(MAX_ROWS)
+        .get_results(conn)
+}
+
+#[derive(Debug, Clone)]
+struct ExpiryNotificationEmail<'a> {
+    name: &'a str,
+    token_name: &'a str,
+    expiry_date: chrono::DateTime<chrono::Utc>,
+}
+
+impl<'a> Email for ExpiryNotificationEmail<'a> {
+    const SUBJECT: &'static str = "Your token is about to expire";
+
+    fn body(&self) -> String {
+        format!(
+            r#"Hi {},
+
+We noticed your token "{}" will expire on {}.
```

Comment:
one thing I just noticed: I don't think we are restricting the token names in any way. technically someone could put e.g. HTML in the token name resulting in a potential security risk when we put it directly in the email. we are using `text/plain` though, so unless some email client ignores that content type it would probably be fine.

since the email will only be sent to the person that created the token this shouldn't be an issue in practice anyway, but it's something to look out for in other generated emails :)